### ISSUE TITLE
FR: Choosable font and time format options for status bar

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#11011]
**Does your feature request involve difficulty completing a task? Please describe.**
Currently, there is no setting of time format for Status bar (bottom). 

**Describe the solution you'd like**
I would like to have an option for time with leading zero, also the font customizable (so I can choose monospace one) but only for time in the status bar. E.g., “09:08” instead of “9:08”.

The reason is because the status bar shifts when minutes are updated. This distracts me a little during reading, also the whole status bar is slight becomes fuzzier than refreshed anew. Having a leading zero and monospace font will solve these issues. 

**Describe alternatives you've considered**
For now I just removed time from status bar and check time only in Top menu. 

**Additional context**
In top menu settings - Device - Date and time, there is an option to choose time format. How ever, I believe this only applies for stats menus like readings stats etc, since status bar and top menu time don’t seem to be affected by this setting. 

This menu should probably clarify this. Also I don’t think one setting should be applied for any time indicators, so this menu’s function is OK as it is.

Thanks!

**Comment by hius07** [comment]
You can test monospace.
https://github.com/koreader/koreader/blob/e9051353a20e5017ed9dddd914ab9a242b45b6a1/frontend/apps/reader/modules/readerfooter.lua#L436
Change to "scfont". Does it look good?

**Comment by mergen3107** [comment]
If this makes the whole footer text monospace, then this wouldn’t work for me. I’d like to have chapter text as usual UI font (that I also changed with a patch). 

(I don’t have access to a laptop right now, but somehow I can’t edit the file from Android in Total Commander, I keep getting access denied. I am sure I could do edit lua files before).

**Comment by hius07** [comment]
Yes, the footer is shown as one string. I doubt this to be changed.

**Comment by mergen3107** [comment]
Alright :) 
Thanks @hius07 !
I will close this FR, it is super niche anyway :D

**Comment by mergen3107** [comment]
Alright, I made it to my laptop! 
Updates!
1) Thanks @hius07 , scfont selected the monospace font. However, which monospace font is this? It is not the one I selected by default in Top menu - Fonts. I think this one is for books only. Can I set it to “Consolas” (font) explicitly?
2) I found in the same file references to datetime. In datetime.lua in the function datetime.secondsToHours for non-Windows OS (what is the Windows os by the way? Lua references?), I replaced “%-H” by just “%H”. Works so far!

**Comment by NiLuJe** [comment]
1. c.f., https://github.com/koreader/koreader/blob/2ed2c2c23da15b9335f7a664fa5cf35f55b1523b/frontend/ui/font.lua#L59-L60

That data structure ought to be easily user-patcheable ;).

**Comment by mergen3107** [comment]
> 1. Can I set it to “Consolas” (font) explicitly?

It turns out, yes! It works, however I don’t know what’s going to happen if Consolas font dies for any reason :D 

Moving on, can I make these two changes as a patch? I think this would be easier than making a fourth argument in the function (named “leading_zeros”, for example), because I imagine I’d need to fix many datetime.secondsToHour call in the code…

**Comment by mergen3107** [comment]
Thanks @NiLuJe !
Missed your comment while typing my comments above :D 

So, could you please help me compile a patch which:
1) makes footer use “scfont”
2) font.lua make scfont Consolas
3) datetime.secondsToHour changes ?

**Comment by mergen3107** [comment]
By the way, I think I found out why I couldn’t modify the files from Android. I am doing it from WinSCP now, and new thing is that some files (like readerfooter.lua) change owner and group from “u0_a153” to “root[0]”. If I modify file and restrat KOReader, it crashes. If I then reset the owner and group to this a153, it works! This wasn’t the case before, IIRC.

**Comment by mergen3107** [comment]
Here is the progress so far.

2-footer-time-zeros-monospace.lua
```
local ReaderFooter = require("apps/reader/modules/readerfooter")
local Font = require("ui/font")
local datetime = require("datetime")

for k, v in pairs(ReaderFooter) do
    if v == "ffont" then
        ReaderFooter[k] = "scfont"
    end
end

for k, v in pairs(Font.fontmap) do
    if v == "DroidSansMono.ttf" then
        Font.fontmap[k] = "Consolas.ttf"
    end
end
```

`datetime.lua` replacements are outside my understanding :D How should I do this safely?

I just need two replacements, basically:
`return os.date(_("%-I:%M AM"), seconds)` to `return os.date(_("%I:%M AM"), seconds)` and
`return os.date(_("%-I:%M PM"), seconds)` to `return os.date(_("%I:%M PM"), seconds)`.

**Comment by mergen3107** [comment]
The font and footer parts work.

**Comment by hius07** [comment]
> for k, v in pairs(ReaderFooter) do
>     if v == "ffont" then
>         ReaderFooter[k] = "scfont"
>     end
> end

ReaderFooter.text_font_face = "scfont"

**Comment by mergen3107** [comment]
While I was testing the time, I noticed that when the Top menu is open, the clock does not update every minute. Is this intentional?

**Comment by hius07** [comment]
> when the Top menu is open, the clock does not update every minute

https://github.com/koreader/koreader/blob/2ed2c2c23da15b9335f7a664fa5cf35f55b1523b/frontend/apps/reader/modules/readerfooter.lua#L793-L796

**Comment by mergen3107** [comment]
Sorry, I wasn't clear. The footer clock updates OK, it is the Top menu clock that doesn't.

**Comment by hius07** [comment]
IIRC, there was a discussion and all agreed that it was weird to stare at the menu longer than a minute.

**Comment by poire-z** [comment]
^ See #9796. It ended up being one of our first user patch: https://gist.github.com/zwim/40ab3866bee35f414feed95d1717e4c1

> The reason is because the status bar shifts when minutes are updated. This distracts me a little during reading, also the whole status bar is slight becomes fuzzier than refreshed anew. Having a leading zero and monospace font will solve these issues.

Don't you have the same issue everytime some page number/page left crosses from 9 to 10 ? (Which must happens most often than crossing 9:59 to 10:00.)

**Comment by offset-torque** [comment]
I think this feature might be useful when it is exposed in the proper UI. See, we had discussions about utilizing the alternative status bar or making this status bar 2 lines to increase the information density.

This way, users can choose for example:
- a "condensed font that is designed to be legible at small sizes" to maximize shown characters if they have "chapter titles" shown for example
- or a "display optimized wider font with lower height" to make the status bar more legible without losing much bottom margin

I accept that, this is kind of an advanced user feature and most of our users wont change anything but on the other hand we already have all the components (status bar menu and font selection menu) and there are reader apps like Moon Reader which exposes this option to the users for a very long time so I am throwing the idea here.

**Comment by mergen3107** [comment]
> Don't you have the same issue everytime some page number/page left crosses from 9 to 10 ? 

This might be my next target, yes 😅
However, there I'd need to calculate how many leading zeros I need. For example, 
01/100 vs 001/100…

**Comment by hius07** [comment]
> how many leading zeros I need

To have the same number of digits as in total pages number.

**Comment by poire-z** [comment]
And the nb of highlights/notes crossing from 9 to 10.... Many items actually.
Come on, live with it :)

**Comment by mergen3107** [comment]
@poire-z 
Thanks for the patch! It restored my inner peace.



> To have the same number of digits as in total pages number.

Yes, that's I meant, but I am not sure I _know_ how to calculate this yet.



> And the nb of highlights/notes crossing from 9 to 10.... Many items actually.

I had to start somewhere :D (and I don't use other parameters yet, I might another ticket when I will xD)

**Comment by mergen3107** [comment]
So for page numbers, in the `rederfooter.lua` function named `page_progress = function(footer)` I changed
```
return ("%d / %d"):format(footer.pageno, footer.pages)
```
to
```
return ("%03d / %d"):format(footer.pageno, footer.pages)
```
This adds leading zeros up to three digits. I am happy for now :D but this change won't survive the update.

**Comment by mergen3107** [comment]
Here is how it looks like in the end:

![Screenshot_20231015-095944](https://github.com/koreader/koreader/assets/6279855/1e4e8886-b07e-4804-ace7-903d4263e49b)

Works absolutely stunning, no more distractions in the footer!

**Comment by mergen3107** [comment]
While testing I noticed that KOReader is about 5-6 seconds behind the time set in Android.
What time does KOReader use? Can I sync it in seconds as well?

The way I set the time is in Android Settings - Date and Time - Use network provided time, toggle off and on.
This syncs the Android top bar clock with my laptop clock, but KOReader updates the minute in footer and top menu clocks 5-6 seconds later.

**Comment by poire-z** [comment]
> but KOReader updates the minute in footer and top menu clocks 5-6 seconds later.

You should know by now that we are really clever - and that we discovered (and may be we didn't share this with the world) that minutes actually change only every 60 seconds ! So, we made KOReader update the footer and issue eInk refreshes only once every 60 seconds, because it's not needed until then :)
So, if you start KOReader at 16:25:32, it shows 16:25 - and it will only show 16:26 at 16:26:32.
(Of course, we didn't think about people like you, and Android where refreshes are cheap, at the time :)

**Comment by mergen3107** [comment]
Makes sense now! Thank you :) 

I was just afraid that this another Luajit-on-Android overhead/shenanigan. 

(Not that I was complaining, I just reported what I see and decided to ask you whether this was intentional)

**Comment by Frenzie** [comment]
> Android where refreshes are cheap, at the time :)

Eh, plenty of people using Android with E Ink screens. ;-)

**Comment by NiLuJe** [comment]
The footer should actually pretty specifically *always* refresh on :01, and not simply every 60s ;).

https://github.com/koreader/koreader/blob/2ed2c2c23da15b9335f7a664fa5cf35f55b1523b/frontend/apps/reader/modules/readerfooter.lua#L820

**Comment by mergen3107** [comment]
That doesn't seem to be the case. 
I'll try to get verbose logs later.

**Comment by mergen3107** [comment]
Footer and top menu clocks change the minute exactly at the same time though

**Comment by Frenzie** [comment]
It does seem to be always on the minute.

**Comment by mergen3107** [comment]
I don't know anything anymore. I was testing it with and without the patch for top menu. Clocks in the status bar and top menu have been anything but consistent. 
1) Now there is no 6 second delay anymore, they update quite well spot on at the 0 second of the next minute, synced with my phone.
2) However, this is only the case for the first update. It seems that KOReader (or my ereader) go to suspend and clocks don't update unless I tap, swipe or press back button. Are the clocks supposed to keep updating, or is there a scheduled suspend on KOReader side? If not, I'll have to go and see Android/Onyx suspend settings.

**Comment by NiLuJe** [comment]
A system standby would certainly explain it ;). And, no, we can't do that ourselves on Android, so it's definitely the ROM ;).

**Comment by mergen3107** [comment]
@NiLuJe 
I tested this standby issue in a different reader. There it updates correctly. 

Can you please show me where to check whether the "Auto-update" option for the status bar works? I have suspicion it broke somehow. I am technically on 2023.08 + 1 patch for History and 1 patch for fonts in status bar.

**Comment by NiLuJe** [comment]
On which device?

We'll wakeup from standby for the next scheduled task (e.g., the footer refresh) on the very few platforms where we implement standby ourselves (i.e., Kobo, and that's it ;p).

That's not the case anywhere else (not even on PocketBook, where we currently defer to the native standby handling); and won't *ever* be the case on platforms where we're not even aware of the fact that the system could standby behind our back (i.e., Android) ;).

----

It's a simple scheduled task, I linked to it above, there's a near zero chance it "broke", what's in play here is standby behavior, that's it ;).

**Comment by mergen3107** [comment]
Alright, thanks :D 
Yes, I meant Android. 

Most probably, status bar updated the same way (until standby) all this time since Auto-refresh feature was added. I just didn't test it more than a minute :D

**Comment by mergen3107** [comment]
Thanks everyone for your help and patience!
I decided not to use the leading zeros after all. Reasoning is that my main distraction was when I was reading a page, closer to the last few lines, and time got updated so that whole reader footer shifted.
Leading zeros before hour would help only when 12:59 becomes 1:00, and 9:59 becomes 10:00, so only four times per 24-hours.
Leading zeros before pages are useless, because they only happen when turn the page myself, i.e., I am not reading at that exact moment anyway :D 

Here is the user patch I used and tested (and keep using): 
[2-footer-time-zeros-monospace.lua.txt](https://github.com/koreader/koreader/files/13062126/2-footer-time-zeros-monospace.lua.txt)
It does make the footer font monospace.
Optionally, it replaces the monospace font to Consolas. This one is commented out right now, please enable it if needed.
